hadoop-hdds/docs/content/design/lifecycle-configurations.md

```diff
@@ -0,0 +1,301 @@
+```protobuf
+/**
+S3 lifecycles (filter, expiration, rule and configuration).
+ */
+message LifecycleFilter {
+  optional string prefix = 1;
+}
+
+message LifecycleExpiration {
+  optional uint32 days = 1;
+  optional string date = 2;
+}
+
+message LifecycleRule {
+  optional string id = 1;
+  optional string prefix = 2;
+  required bool enabled = 3;
+  optional LifecycleExpiration expiration = 4;
+  optional LifecycleFilter filter = 5;
+}
+
+message LifecycleConfiguration {
+  required string volume = 1;
+  required string bucket = 2;
+  required string owner = 3;
+  optional uint64 creationTime = 4;
+  repeated LifecycleRule rules = 5;
+  optional uint64 objectID = 6;
+  optional uint64 updateID = 7;
+}
+
+message CreateLifecycleConfigurationRequest {
+  required LifecycleConfiguration lifecycleConfiguration = 1;
+}
+
+message CreateLifecycleConfigurationResponse {
+
+}
+
+message InfoLifecycleConfigurationRequest {
+  required string volumeName = 1;
+  required string bucketName = 2;
+}
+
+message InfoLifecycleConfigurationResponse {
+  required LifecycleConfiguration lifecycleConfiguration = 1;
+}
+
+message DeleteLifecycleConfigurationRequest {
+  required string volumeName = 1;
+  required string bucketName = 2;
+}
+
+message DeleteLifecycleConfigurationResponse {
+
+}
+
+message ListLifecycleConfigurationsRequest {
+  optional string userName = 1;
+  optional string prevKey = 2;
+  optional uint32 maxKeys = 3;
+}
+
+message ListLifecycleConfigurationsResponse {
+  repeated LifecycleConfiguration lifecycleConfiguration = 1;
+}
+```
+
+# Table format
+## OmLifecycleConfiguration Table
+
+The `OmLifecycleConfiguration` table in RocksDB is used to store lifecycle configurations of buckets. Below is a summary of the table structure.
+
+### Table Structure
+
+| Column Name  | Data Type             | Description                                              |
+|--------------|-----------------------|----------------------------------------------------------|
+| volume       | String                | The name of the volume.                                  |
+| bucket       | String                | The name of the bucket.                                  |
+| owner        | String                | The owner of the volume/bucket.                          |
+| creationTime | long                  | The creation time of the configuration.                  |
+| rules        | List<OmLCRule>        | A list of lifecycle rules associated with the configuration. |
+| objectID     | long                  | Unique identifier for the object.                        |
+| updateID     | long                  | Identifier for updates to the object.                    |
+
+
+### Additional Information
+
+- **Maximum Rules**: The table can store up to 1000 rules per lifecycle configuration.
+- **Validation**: The configuration is considered valid if:
+  - The `volume`, `bucket`, and `owner` are not blank.
+  - The number of rules is between 1 and 1000.
```

Comment:
Do we plan on introducing a property for setting the number of rules to be stored per lifecycle configuration?

Reply:
Not really, each lifecycle configuration will have a list of rules. Hence there is no need to explicitly maintain the count of rules.

Reply:
We can limit it to 1000 rules per lifecycle configuration as per AWS documentation (https://docs.aws.amazon.com/AmazonS3/latest/userguide/intro-lifecycle-rules.html#intro-lifecycle-rule-id).

Reply:
Understood, thank you for clarifying @ivandika3 and @mohan3d.

Reply:
@mohan3d Please help to add this in the design docs.

Reply:
@ivandika3 you mean the implementation details on how it will be limited to 1000 rules or the limit itself? the limit has been added earlier.

Reply:
Yes I saw that it's been specified. Probably better to add the link the actual AWS documentation so the readers does not misconstrue it as an arbitrary constraints. 

The other AWS docs links can also be included in other parts of the design documents (e.g. conflicting rules).